namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala

```diff
@@ -329,50 +311,69 @@ class ThriftNamerInterface(
-            getAddrObserver(path) match {
-              case Some(observer) =>
-                respondAddr(observer, reqStamp)
-
-              case None =>
-                val e = thrift.AddrFailure(s"no bound address for ${path.show}", -1, ref)
-                Future.exception(e)
-            }
+    val thrift.AddrReq(ref@thrift.NameRef(reqStamp, reqName, _), _) = req
+    mkPath(reqName) match {
+      case Path.empty =>
+        Trace.recordBinary("namerd.srv/addr.err", "empty path")
+        val failure = thrift.AddrFailure("empty path", Int.MaxValue, ref)
+        Future.exception(failure)
 
-          case Throw(e) =>
-            val msg = s"error binding ${path.show}: ${e.getMessage}"
-            Future.exception(thrift.AddrFailure(msg, retryIn().inSeconds, ref))
+      case path =>
+        Trace.recordBinary("namerd.srv/addr.path", path.show)
+        val addrObserver = observeAddr(path)
+        addrObserver(reqStamp).map {
+          case (newStamp, None) =>
+            Trace.recordBinary("namerd.srv/addr.result", "neg")
+            thrift.Addr(TStamp(newStamp), thrift.AddrVal.Neg(TVoid))
+
+          case (newStamp, Some(bound@Addr.Bound(addrs, meta))) =>
+            Trace.recordBinary("namerd.srv/addr.result", bound.toString)
+            val taddrs = addrs.collect {
+              case Address.Inet(isa, _) =>
+                // TODO translate metadata (weight info, latency compensation, etc)
+                val ip = ByteBuffer.wrap(isa.getAddress.getAddress)
+                thrift.TransportAddress(ip, isa.getPort)
+            }
+            thrift.Addr(TStamp(newStamp), thrift.AddrVal.Bound(thrift.BoundAddr(taddrs)))
         }
     }
   }
 
-  private[this] def respondAddr(observer: AddrObserver, reqStamp: TStamp): Future[thrift.Addr] = {
-    observer(reqStamp).map {
-      case (newStamp, None) =>
-        thrift.Addr(TStamp(newStamp), thrift.AddrVal.Neg(TVoid))
-
-      case (newStamp, Some(Addr.Bound(addrs, meta))) =>
-        val taddrs = addrs.collect {
-          case Address.Inet(isa, _) =>
-            // TODO translate metadata (weight info, latency compensation, etc)
-            val ip = ByteBuffer.wrap(isa.getAddress.getAddress)
-            thrift.TransportAddress(ip, isa.getPort)
-        }
-        thrift.Addr(TStamp(newStamp), thrift.AddrVal.Bound(thrift.BoundAddr(taddrs)))
+  private[this] val addrCacheMu = new {}
+  private[this] var addrCache: Map[Path, AddrObserver] = Map.empty
+  private[this] def observeAddr(id: Path): AddrObserver =
+    addrCacheMu.synchronized {
+      addrCache.get(id) match {
+        case Some(obs) =>
+          Trace.recordBinary("namerd.srv/addr.cached", true)
+          obs
+
+        case None =>
+          Trace.recordBinary("namerd.srv/addr.cached", false)
+          val (pfx, namer) = namers.find { case (p, _) => id.startsWith(p) }.getOrElse(DefaultNamer)
+          val resolution = namer.bind(NameTree.Leaf(id.drop(pfx.size))).run.flatMap {
```

Comment:
sorry, I should have been more specific.  I meant to possibly factor out the `Activity.State => Var[Resolution]` mapping since it's long and interrupts the flow.

Not a big deal, TIOLI.